tests/ui/feature-gates/feature-gate-pin_ergonomics.rs

```diff
@@ -1,15 +1,16 @@
-#![allow(dead_code, incomplete_features)]
+#![allow(dead_code)]
 
 use std::pin::Pin;
 
 struct Foo;
 
-fn foo(_: Pin<&mut Foo>) {
-}
+fn foo(_: &pin mut Foo) {} //~ ERROR pinned reference syntax is experimental
 
-fn bar(mut x: Pin<&mut Foo>) {
+fn bar(x: Pin<&mut Foo>) {
     foo(x);
     foo(x); //~ ERROR use of moved value: `x`
 }
 
+fn baz(_: &pin const Foo) {} //~ ERROR pinned reference syntax is experimental
```

Comment:
What will happen if we had a macro on stable that has a type fragment in the first macro matcher and a token tree in the second. Would we now pick the first and error or sth if used as `themacro!(&pin const T)`

Reply:
Arbitrary token trees (nor `mut` nor `const`) are in the follow set for a type fragment, so this shouldn't happen afaict?

Reply:
```rust
macro_rules! foo {
    ($x:ty) => { compile_error!("bad") };
    ($($t:tt)*) => {}
}

foo!(&pin mut u32);
```

Compiles today, but may be affected by this PR either directly or just if the feature is enabled

Reply:
Yeah, though I'd consider this to be a theoretical breakage, similarly to adding `async` and `const` to trait bounds.

Reply:
Agreed, but it should probably not break while the feature is still unstable I think. Only at stabilization time

Reply:
I'm not certain if there's a feasible way to prevent this breakage without simply excluding `&pin mut T` from the type nonterminal, which seems not good

Reply:
I think this is covered by [RFC 3531](https://rust-lang.github.io/rfcs/3531-macro-fragment-policy.html). Basically, until the edition after this is hypothetically stabilized, `$ty` continues to match what it does currently. Then for the next edition, we'd update `$ty` to include the new syntax and add `$ty_2024` or some other more descriptive name to keep the existing behavior.

Reply:
That's a lot of complexity. I agree that this is a highly artificial situation and will not occur in practice. If it had been easy to avoid we could have done so, but it isn't, so let's go ahead without any additional checks

Reply:
Do you think it's worth adding your example as a test case to make sure we don't accidentally change the behavior of the `ty` matcher? I'd be happy to add it, although I think the matcher implementation is surprisingly separate from the parser so I'm not sure there's a huge risk of accidentally doing this.

Reply:
I don't think it's needed, but don't feel strongly either way.

Someone from @rust-lang/lang should probably look at this and decide if it's worthwile discussing or even FCPing.